java/fury-core/src/main/java/org/apache/fury/reflect/TypeRef.java

```diff
@@ -0,0 +1,828 @@
+    for (Type bound : bounds) {
+      final Type componentType = getComponentType(bound);
+      if (componentType != null) {
+        // Only the first bound can be a class or array.
+        // Bounds after the first can only be interfaces.
+        if (componentType instanceof Class) {
+          final Class<?> componentClass = (Class<?>) componentType;
+          if (componentClass.isPrimitive()) {
+            return componentClass;
+          }
+        }
+        return componentType;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Resolves the given {@code type} against the type context represented by this type. For example:
+   *
+   * <pre>{@code
+   * new TypeToken<List<String>>() {}.resolveType(
+   *     List.class.getMethod("get", int.class).getGenericReturnType())
+   * => String.class
+   * }</pre>
+   */
+  public TypeRef<?> resolveType(Type iteratorReturnType) {
+    if (iteratorReturnType instanceof WildcardType) { // fast path
+      return of(iteratorReturnType);
+    }
+    Type invariantContext = WildcardCapturer.capture(type);
+    Map<Types.TypeVariableKey, Type> mappings = resolveTypeMappings(invariantContext);
+    return resolveType0(iteratorReturnType, mappings);
+  }
+
+  private TypeRef<?> resolveType0(
+      Type iteratorReturnType, Map<Types.TypeVariableKey, Type> mappings) {
+    if (iteratorReturnType instanceof TypeVariable) {
+      TypeVariable<?> typeVariable = (TypeVariable<?>) iteratorReturnType;
+      Type type = mappings.get(new Types.TypeVariableKey(typeVariable));
+      if (type == null) {
+        return of(typeVariable);
+      }
+      return resolveType0(type, mappings);
+    } else if (iteratorReturnType instanceof ParameterizedType) {
+      ParameterizedType parameterizedType = (ParameterizedType) iteratorReturnType;
+      Type owner = parameterizedType.getOwnerType();
+      Type resolvedOwner = owner == null ? null : resolveType0(owner, mappings).type;
+      Type resolvedRawType = resolveType0(parameterizedType.getRawType(), mappings).type;
+
+      Type[] args = parameterizedType.getActualTypeArguments();
+
+      Type[] resolvedArgs = new Type[args.length];
+      for (int i = 0; i < args.length; i++) {
+        resolvedArgs[i] = resolveType0(args[i], mappings).type;
+      }
+
+      return of(new Types.ParameterizedTypeImpl(resolvedOwner, resolvedRawType, resolvedArgs));
+    } else if (iteratorReturnType instanceof GenericArrayType) {
+      Type componentType = ((GenericArrayType) iteratorReturnType).getGenericComponentType();
+      Type resolvedComponentType = resolveType0(componentType, mappings).type;
+      return of(newArrayType(resolvedComponentType));
+    }
+    return of(iteratorReturnType);
+  }
+
+  private static Map<Types.TypeVariableKey, Type> resolveTypeMappings(Type contextType) {
+    Map<Types.TypeVariableKey, Type> result = new HashMap<>();
+    populateTypeMapping(result, contextType);
+    return result;
+  }
```

Comment:
@chaokunyang this method was also cached in Guava's TypeToken. What do you think about this? I believe it might be worth to add caching for `contextType = type` parameter, but not for invariant type.

Reply:
Yep, creating map is costly, cache is better. But be careful do not hold strong reference, otherwise the class is nog eligible for GC

Reply:
I'm not sure how guava aovid class gc issue for such cache. Could you share some details? Maybe we can use it to optimize fury class cache too

Reply:
I'm not sure what do you mean by "strong reference" here.
If you are talking about `Type contextType` argument - we wont cache it at all, since this argument is always equals to either `this.type` or `WildcardCapturer.capture(this.type)`, so we will cache the result of this method when `contextType = this.type`.
Guava didn't use weak references here, they are just storing this data in `ImmutableMap`.

Reply:
If we cache it in a static map field, when the data in this map got collected by java gc?

Reply:
Not static, but TypeRef's local field.

Reply:
Caching this in static field is an interesting idea though. We can use som weak map for this, but I think that it is better to implement this in a separate PR, since this PR already became too big.

Reply:
I see, I think we can cache as an instance field in this PR